front50-s3/src/main/java/com/netflix/spinnaker/front50/model/S3StorageService.java

```diff
@@ -0,0 +1,253 @@
+
+  private final ObjectMapper objectMapper;
+  private final AmazonS3 amazonS3;
+  private final String bucket;
+  private final String rootFolder;
+
+  public S3StorageService(ObjectMapper objectMapper,
+                          AmazonS3 amazonS3,
+                          String bucket,
+                          String rootFolder) {
+    this.objectMapper = objectMapper;
+    this.amazonS3 = amazonS3;
+    this.bucket = bucket;
+    this.rootFolder = rootFolder;
+  }
+
+  @Override
+  public void ensureBucketExists() {
+    // do nothing
+  }
+
+  @Override
+  public boolean supportsVersioning() {
+    return true;
+  }
+
+  @Override
+  public <T extends Timestamped> Collection<T> loadObjectsWithPrefix(ObjectType objectType, String prefix, int maxResults) {
+    ObjectListing bucketListing = amazonS3.listObjects(
+      new ListObjectsRequest(bucket, (buildTypedFolder(rootFolder, objectType.group) + "/" + prefix).toLowerCase(), null, null, maxResults)
+    );
+    List<S3ObjectSummary> summaries = bucketListing.getObjectSummaries();
+
+    // TODO-AJ this is naive and inefficient
+    return summaries.stream()
+      .map(s3ObjectSummary -> amazonS3.getObject(s3ObjectSummary.getBucketName(), s3ObjectSummary.getKey()))
+      .map(s3Object -> {
+        T item = null;
+        try {
+          item = deserialize(s3Object, (Class<T>) objectType.clazz);
+          item.setLastModified(s3Object.getObjectMetadata().getLastModified().getTime());
+        } catch (IOException e) {
+          // do nothing
+        }
+        return item;
+      })
+      .collect(Collectors.toSet());
+  }
+
+  @Override
+  public <T extends Timestamped> T loadObject(ObjectType objectType, String objectKey) throws NotFoundException {
+    try {
+      S3Object s3Object = amazonS3.getObject(bucket, buildS3Key(objectType.group, objectKey, objectType.defaultMetadataFilename));
+      T item = deserialize(s3Object, (Class<T>) objectType.clazz);
+      item.setLastModified(s3Object.getObjectMetadata().getLastModified().getTime());
+      return item;
+    } catch (AmazonS3Exception e) {
+      if (e.getStatusCode() == 404) {
+        throw new NotFoundException("Object not found (key: " + objectKey + ")");
+      }
+      throw e;
+    } catch (IOException e) {
+      throw new IllegalStateException("Unable to deserialize object (key: " + objectKey + ")", e);
+    }
+  }
+
+  @Override
+  public <T extends Timestamped> T loadObjectVersion(ObjectType objectType,
+                                                     String objectKey,
+                                                     String versionId) throws NotFoundException {
+    throw new UnsupportedOperationException("loadObjectVersion() is not supported");
+  }
+
+  @Override
+  public void deleteObject(ObjectType objectType, String objectKey) {
+    amazonS3.deleteObject(bucket, buildS3Key(objectType.group, objectKey, objectType.defaultMetadataFilename));
```

Comment:
In create you wrap errors but not here.
I chose to ignore 404. We should be consistent, so I assume you'd want the same. If not I'll take mine out. Unless you wish you could 404 but cannot for legacy reasons.